### PR TITLE
refactor: Bump OpenDAL 0.46, arrow 51, tonic 0.11, reqwest 0.12, hyper 1, http 1

### PR DESCRIPTION
**Pull request by Xuanwo**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

This PR will bump the following pacakges:

- OpenDAL 0.46
- arrow 51
- tonic 0.11
- reqwest 0.12
- hyper 1
- http 1

I have to submit them all in one PR to make sure databend can build.

## Changes that worth of mention

OpenDAL v0.46 has transitioned from AsyncRead-based IO to Range-based IO. While some internal implementations have been refactored, not all are performing optimally. We need to address this regression.

We have transitioned to range-based IO for both Parquet and native formats; please give this area extra attention.

To minimize the number of changes in the PR, I refactored directly without robust abstraction, such as using opendal::Reader directly in the native format API. I intend to refine this aspect in upcoming PRs.

## Benchmarks

Concurrent reading is enabled only for loading CSV/TSV files; other queries remain unaffected. I will first ensure there are no regressions, and then I'll optimize Parquet reading.

### Qeury Performance 

hits

image

tpch

image

### Load (with 2 concurrent)

image

### Load (with 4 concurrent)

image

## Tests

- [x] Unit Test
- [x] Logic Test
- [x] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [ ] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [x] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15442)

### CODE REVIEW

**Comment by Xuanwo** [comment]
Build failed for:

```rust
error[E0658]: use of unstable library feature 'stdsimd'
  --> /opt/rust/cargo/registry/src/index.crates.io-6f17d22bba15001f/crc32c-0.6.5/src/hw_aarch64.rs:50:33
   |
50 |         .fold(crc, |crc, &next| simd::__crc32cb(crc, next))
   |                                 ^^^^^^^^^^^^^^^
   |
   = note: see issue #48556 <https://github.com/rust-lang/rust/issues/48556> for more information
   = help: add `#![feature(stdsimd)]` to the crate attributes to enable
   = note: this compiler was built on 2024-01-22; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'stdsimd'
  --> /opt/rust/cargo/registry/src/index.crates.io-6f17d22bba15001f/crc32c-0.6.5/src/hw_aarch64.rs:62:5
   |
62 |     simd::__crc32cd(crc, next)
   |     ^^^^^^^^^^^^^^^
   |
   = note: see issue #48556 <https://github.com/rust-lang/rust/issues/48556> for more information
   = help: add `#![feature(stdsimd)]` to the crate attributes to enable
   = note: this compiler was built on 2024-01-22; consider upgrading it if it is out of date

error[E0635]: unknown feature `stdarch_arm_crc32`
  --> /opt/rust/cargo/registry/src/index.crates.io-6f17d22bba15001f/crc32c-0.6.5/src/lib.rs:23:60
   |
23 | #![cfg_attr(all(target_arch = "aarch64", nightly), feature(stdarch_arm_crc32))]
   |                                                            ^^^^^^^^^^^^^^^^^

Some errors have detailed explanations: E0635, E0658.
For more information about an error, try `rustc --explain E0635`.
error: could not compile `crc32c` (lib) due to 3 previous errors
```

I'm working on the fix.

**Review by sundy-li — APPROVED** [review]

**Comment by Xuanwo** [comment]
I'm waiting for the benchmark results.

**Comment by Xuanwo** [comment]
The benchmark seems quiet slow, I'm working on this.

File: src/query/storages/parquet/src/parquet_rs/parquet_reader/reader/full_reader.rs [hunk 1]
```diff
@@ -73,13 +80,15 @@ impl ParquetRSFullReader {
...
         loc: &str,
         partition_fields: Option<&[(TableField, Scalar)]>,
-    ) -> Result<ParquetRecordBatchStream<Reader>> {
+    ) -> Result<ParquetRecordBatchStream<ParquetFileReader>> {
         let partition_values_map = partition_fields.map(|arr| {
             arr.iter()
                 .map(|(f, v)| (f.name().to_string(), v.clone()))
                 .collect::<std::collections::HashMap<String, Scalar>>()
         });
+        let meta = self.op.stat(loc).await?;
```
Comment (youngsofun) [inline src/query/storages/parquet/src/parquet_rs/parquet_reader/reader/full_reader.rs:89]: this meta is available when listing files?

Comment (Xuanwo) [reply src/query/storages/parquet/src/parquet_rs/parquet_reader/reader/full_reader.rs:89]: > this meta is available when listing files?

Yes, maybe I can change the code a bit to use existing content length without extra stat call.

File: src/query/storages/stage/src/read/row_based/processors/reader.rs [hunk 2]
```diff
@@ -48,30 +46,31 @@ impl BytesReader {
...
         op: Operator,
-        read_batch_size: usize,
+        _read_batch_size: usize,
         prefetch_num: usize,
     ) -> Result<Self> {
         Ok(Self {
             table_ctx,
             op,
-            read_batch_size,
+            // TODO(xuanwo): bring batch size back when we figure how to play with range based read.
```
Comment (youngsofun) [inline src/query/storages/stage/src/read/row_based/processors/reader.rs:55]: why not use the setting?

File: src/query/storages/stage/src/read/row_based/processors/reader.rs [hunk 3]
```diff
@@ -48,30 +46,31 @@ impl BytesReader {
...
         if let Some(state) = &mut self.file_state {
-            let end = min(self.read_batch_size + state.offset, state.file.size);
-            let mut buffer = vec![0u8; end - state.offset];
-            let n = read_full(&mut state.reader, &mut buffer[..]).await?;
-            if n == 0 {
+            let end = min(self.read_batch_size + state.offset, state.file.size) as u64;
+
+            let buffer = state.reader.read(state.offset as u64..end).await?.to_vec();
+            let n = buffer.len();
+            if (n as u64) < end - state.offset as u64 {
```
Comment (youngsofun) [inline src/query/storages/stage/src/read/row_based/processors/reader.rs:68]: log batch size in the error message if you need?

File: src/query/storages/stage/src/read/row_based/processors/reader.rs (hunk 2 above)
Comment (Xuanwo) [reply src/query/storages/stage/src/read/row_based/processors/reader.rs:55]: Apologies for the confusion; this change is solely for debugging purposes. I have a new implementation locally and will invite you to review it once I've resolved the benchmark issues.

File: src/query/storages/stage/src/read/row_based/processors/reader.rs (hunk 3 above)
Comment (Xuanwo) [reply src/query/storages/stage/src/read/row_based/processors/reader.rs:68]: https://github.com/datafuselabs/databend/pull/15442#discussion_r1596175494

File: src/query/storages/stage/src/read/row_based/processors/reader.rs [hunk 4]
```diff
@@ -48,30 +46,31 @@ impl BytesReader {
...
 
     pub async fn read_batch(&mut self) -> Result<DataBlock> {
         if let Some(state) = &mut self.file_state {
-            let end = min(self.read_batch_size + state.offset, state.file.size);
-            let mut buffer = vec![0u8; end - state.offset];
-            let n = read_full(&mut state.reader, &mut buffer[..]).await?;
-            if n == 0 {
+            let end = min(self.read_batch_size + state.offset, state.file.size) as u64;
+
+            let buffer = state.reader.read(state.offset as u64..end).await?.to_vec();
```
Comment (youngsofun) [inline src/query/storages/stage/src/read/row_based/processors/reader.rs:66]: I used to read like this mistakingly.
FYI, change bache size to 16M is not helpful.

**Comment by Xuanwo** [comment]
Hi, @youngsofun, this PR is ready for review now!

**Comment by BohuTANG** [comment]
image
The data load times improved significantly in OpenDAL 0.46, but query performance as measured by the hits benchmark seems unchanged. What specific optimizations did 0.46 include that sped up loading but not queries?

**Comment by Xuanwo** [comment]
> What specific optimizations did 0.46 include that sped up loading but not queries?

OpenDAL v0.46 introduces a new feature known as concurrent read, also referred to as auto ranged read.
I've only enabled concurrency in a few continuous reading areas like the bytes reader. Other areas remain unchanged, and I'm satisfied with the current performance since there's no significant regression. I plan to optimize additional areas in upcoming PRs.

File: src/query/storages/parquet/src/parquet_rs/parquet_reader/reader/full_reader.rs (hunk 1 above)
Comment (Xuanwo) [reply src/query/storages/parquet/src/parquet_rs/parquet_reader/reader/full_reader.rs]: Fixed by reuse content length in file parts.

**Review by youngsofun — APPROVED** [review]

**Pull request merged**